packages/angular/src/utils/mfe/with-module-federation.ts

```diff
@@ -0,0 +1,238 @@
+import {
+  SharedLibraryConfig,
+  sharePackages,
+  shareWorkspaceLibraries,
+} from './mfe-webpack';
+import {
+  createProjectGraphAsync,
+  readCachedProjectGraph,
+} from '@nrwl/workspace/src/core/project-graph';
+import { readWorkspaceJson } from '@nrwl/workspace/src/core/file-utils';
+import { joinPathFragments, ProjectGraph } from '@nrwl/devkit';
+import ModuleFederationPlugin = require('webpack/lib/container/ModuleFederationPlugin');
+import { Workspaces } from 'nx/src/shared/workspace';
+import { appRootPath } from 'nx/src/utils/app-root';
+import {
+  getRootTsConfigPath,
+  readTsConfig,
+} from '@nrwl/workspace/src/utilities/typescript';
+import { ParsedCommandLine } from 'typescript';
+
+export type MFERemotes = string[] | [remoteName: string, remoteUrl: string][];
+
+export interface MFEConfig {
+  name: string;
+  remotes?: MFERemotes;
+  exposes?: Record<string, string>;
+  shared?: (
+    libraryName: string,
+    library: SharedLibraryConfig
+  ) => SharedLibraryConfig | false;
+}
+
+function recursivelyResolveWorkspaceDependents(
+  projectGraph: ProjectGraph<any>,
+  target: string,
+  collectedDeps: Set<string> = new Set()
+) {
+  if (collectedDeps.has(target)) {
+    return [];
+  }
+  let dependencies = [target];
+
+  const workspaceDependencies = (
+    projectGraph.dependencies[target] ?? []
+  ).filter((dep) => !dep.target.startsWith('npm:'));
+  if (workspaceDependencies.length > 0) {
+    for (const dep of workspaceDependencies) {
+      dependencies = [
+        ...dependencies,
+        ...recursivelyResolveWorkspaceDependents(
+          projectGraph,
+          dep.target,
+          collectedDeps
+        ),
+      ];
+    }
+  }
+  collectedDeps = new Set([...collectedDeps, ...dependencies]);
+  return [...collectedDeps];
+}
```

Comment:
You're creating new instances of the `Set` in every call, which prevents the caller to have access to updated values while the loop is executing because the reference is lost. The idea would be to keep the same reference and mutate the `Set` adding the targets as they get processed so at all times the updated values can be accessed by any call in the stack:

```suggestion
function recursivelyResolveWorkspaceDependents(
  projectGraph: ProjectGraph<any>,
  target: string,
  seenTargets: Set<string> = new Set()
) {
  if (seenTargets.has(target)) {
    return [];
  }
  let dependencies = [target];
  seenTargets.add(target);

  const workspaceDependencies = (
    projectGraph.dependencies[target] ?? []
  ).filter((dep) => !dep.target.startsWith('npm:'));
  if (workspaceDependencies.length > 0) {
    for (const dep of workspaceDependencies) {
      dependencies = [
        ...dependencies,
        ...recursivelyResolveWorkspaceDependents(
          projectGraph,
          dep.target,
          seenTargets
        ),
      ];
    }
  }

  return dependencies;
}
```

Reply:
Oh wow, that diff is useless, it doesn't highlight the actual changes which is not the whole block 🤦🏻

Reply:
Yeah I get what you mean. I misunderstood originally thinking we just wanted to ensure that the return of dependencies was a unique array. 
but it can be better as you suggested.